Improve docs before alpha release (#137)
neuroinformatics-unit/datashuttle: docs/source/pages/documentation.md

# Full Documentation

> DataShuttle is a tool to streamline the management of neuroscience projects, allowing simple and convenient standardisation of data structures.

Reviewer: Do we mean "data structures" here? I'm thinking more "files and folders"?

Reviewer: Also, should the fact that it's a Python and Command Line tool already be mentioned here? Will be useful for potential users to decide early whether `datashuttle` is what they want or not (I know it's mentioned further down)

Author: Yes I think this is a good idea. I'm in a muddle as to where to introduce this, because as you say it is important to know sooner rather than later. However the terms API and CLI are a little technical and might not be immediately obvious even for those from a non-computer science background who know how to code or use a terminal (e.g. I recall not knowing what an API was even though I was reasonably experienced with python).  As such I don't want to introduce it too early and make it seem overly techy.

DataShuttle's goal is to alleviate the burden researchers face in adhering to standardized file and folder specifications during the execution of intricate and demanding experimental projects. It will:

- Eliminate the need to manually integrate datasets collected across different machines (e.g. behaviour and electrophysiology acquisition machines).

> - Allow convenient transfer of data between machines. This may be between a central project storange and analysis machine (e.g. ''*I want to transfer subjects 1-5, sessions 5 and 10, behavioural data only to my laptop*.'')

Reviewer: typo
```suggestion
- Allow convenient transfer of data between machines. This may be between a central project storage and analysis machine (e.g. ''*I want to transfer subjects 1-5, sessions 5 and 10, behavioural data only to my laptop*.'')
```

> - Avoids re-naming and re-formatting of project folders for collaboriation or dataset publication.

Reviewer: typo
```suggestion
- Avoids re-naming and re-formatting of project folders for collaboration or dataset publication.
```

> DataShuttle aims to integreate seamlessly into the  neuroscience data collection and analysis workflows and eliminate the need to manually , providing tools to:

Reviewer: typos
```suggestion
DataShuttle aims to integrate seamlessly into the neuroscience data collection and analysis workflows and eliminate the need to manually, providing tools to:
```

> - Create folder trees that adhere to SWC-Blueprint, a data management specification based on and aligned to the Brain Imaging Dataset Specification (BIDS), widely used in neuroscience.

Reviewer: Should we link to SWC Blueprint docs here?

- Convenient transfer of between machines used for data collection and analysis, and a central storage repository.

[IMAGE OF PCS]

DataShuttle requires a one-time setup of project name and configurations.  Next, subjects, session and data-type folder trees can be conveniently created during experimental acquisition. Once acquisition is complete, data can be easily transferred from acquisition computers to a central storage machine.

### Installation

DataShuttle is hosted on  [PyPI](https://pypi.org/project/datashuttle/) and can be installed with pip.

`pip install datashuttle`

> Datashuttle required Rclone for data transfers. The easiest way to install Rclone is using [Miniconda](https://docs.conda.io/en/main/miniconda.html):

Reviewer: potential typo and suggestion
```suggestion
Datashuttle additionally requires Rclone for data transfers. The easiest way to install Rclone is using [Miniconda](https://docs.conda.io/en/main/miniconda.html):
```

```
conda install -c conda-forge rclone
```

See [the Rclone website](https://rclone.org/install/) for alternative installation methods.

### Getting Started

Datashuttle provides a Python API and cross-platform command line interface (CLI). In this guide examples will be down using the command line, but corresponding methods can be found in the [API Reference](https://datashuttle.neuroinformatics.dev/pages/api_index.html).

The first thing to do when using DataShuttle is to setup a new project on a *local* machine.

#### *local* machines and the *central* machine

DataShuttle makes the distinction between (possibly multiple) *local* machines and a single *central* machine. DataShuttle needs to be setup once for each *local* machine, but requires no setup on the *central* machine.

> A typical use case is an experment in which behavioural data and electrophysiological data are collected on acquisition PCs. They send the data to a central server where it is stored.

Reviewer: typo and small potential clarification
```suggestion
A typical use case is an experiment in which behavioural data and electrophysiological data are collected on (different) acquisition PCs. They send the data to a central server where it is stored.
```

The command `make-config-file` is used for the initial setup of the project. The **required arguments** are:

`local_path`: The full filepath to the project folder on the *local* machine. For example, if you wanted to make a new project called `my_first_project` in the folder `C:\User\my_projects`, the local path would be `C:\User\my_projects`.

`central_path`: The path on the *central* machine to the central project. For example, if connecting to a remote linux server, this may be `/hpc/home/user/my_projects`.

`connection_method`: `local_filesystem` or `ssh`. Local filesystem can be used if the *central* storage is mounted to the local machine. Otherwise `ssh` can be used.

**Optional Arguments**

If connection method is `ssh`, the `central_host_id`, `central_host_username` must be set, and a one-time SSH setup command run (see the [SSH section][#### SSH] for details).

The optional arguments `ovewrite_old_files`, `transfer_verbosity` and `show_transfer_progress` determine how *data transfer* is performed (see the [Data Transfer section](#### Data Transfer) for details).

> Finally, the optional *data-type* flags `--use_ephys`, `--use_funcimg`, `--use_histology`, `--use_behav` set the types of data required for the project on the local machine.

Reviewer: At least one of these must be used though right (from my user testing experiences) - so are these optional?

Author: That's correct thanks, I have moved it to the end of the above section, so now it reads:

```
Finally, the *data-type* flags `--use_ephys`, `--use_funcimg`, `--use_histology`, `--use_behav` set the types of data required for the project on the local machine. While individual flags are optional, at least one must be chosen when initialising the project.

**Optional Arguments**
```

Does this seem clear?

```
datashuttle \
my_first_project \
make-config-file \
/path/to/my/project \
/nfs/nhome/live/username/ \
ssh \
--central_host_id ssh.swc.ucl.ac.uk \
--central_host_username username \
--transfer_verbosity v \
--use-ephys --use-behav --use-histology --overwrite_old_files
```

Now setup is complete! Configuration settings can be edited at any time with the `update-config` command (TODO LINK). Alternatively, custom confiruation files can be supplied using the `supply-config` command (TODO LINK) (this greatly simplifies setting up projects across multiple *local* machines).

Next, we can start setting up the project by automatically creating project folder trees that conform to a standardised specification.

#### Creating *subject* and *session* folders

In a typical neuroscience experiment, a data-collection session begins by creating the folder for the current subject name (e.g. mouse, rat) and current session name. Once created, the data for this session is stored in the created folder.

> The command `make-sub-folders` can be used automatically create folder trees that adhere to the SWC-Blueprint (and correspondingly, BIDS) specification. The linked specifications contain more detail, but at it's heard this requires:

Reviewer: typo
```suggestion
The command `make-sub-folders` can be used automatically create folder trees that adhere to the SWC-Blueprint (and correspondingly, BIDS) specification. The linked specifications contain more detail, but at its heart this requires:
```

All subjects are given a numerical (integer) number that is prefixed with the key `sub-`. All sessions are also given a numerical (integer) number that is prefixed with the key `ses-`.

Following this, optional information can be included in the form of key-value pairs. For example, a folder tree for *subject 1*, *session 1*  with *behavioural* data that includes the date of each session in the *session* folder name would be:

```
└── sub-001/
    └── ses-001_date_20230516/
        └── behav/
            └── sub-001_ses-001_recording.mp4
```

In DataShuttle, this folder tree (excluding the .mp4 file which must be saved using third-party software) can be created (assuming today's date is `20220516`), with the command

```
datashuttle \
my_first_project \
make-sub-folders -sub 001 -ses 001_@DATE@ -dt behav
```

> The leading `sub-` or `ses-` is optional when specifying folders to create. It is possible to automatically create date, time or datetime key-value pairs with the days `@DATE@`, `@TIME@` or `@DATETIME@` repsectively.

Reviewer: ```suggestion
The leading `sub-` or `ses-` is optional when specifying folders to create. It is possible to automatically create date, time or datetime key-value pairs with the days `@DATE@`, `@TIME@` or `@DATETIME@` respectively.
```

Another example call, which creates a range of subject and session folders, is shown below:

```
datashuttle \
my_first_project \
make-sub-folders -sub 001@TO@003 -ses 010_@TIME@ -dt all
```

When the `all` argument is used for `--data_type` (`-dt`), the folders created depend on the *data-types* specified during *configuration* setup. For example, if
`--use_behav`, `--use_funcimg`, `--use_histology` were set during *configuration* setup, the folder tree from the above command (assuming the time is 4.02.48 PM), would look like:

```
├── sub-001/
│   ├── ses-010_time-160248/
│   │   ├── behav
│   │   └── funcimg
│   └── histology
├── sub-002/
│   ├── ses-010_time-160248/
│   │   ├── behav
│   │   └── funcimg
│   └── histology
└── sub-003/
    ├── ses-010_time-160248/
    │   ├── behav
    │   └── funcimg
    └── histology
```

# FOR FULL DETAILS, SEE XXX

#### Data Transfer

#### Configuration

This documentation gives examples both using the API (in the python console) or using the command line interface (in system terminal).

> #### C

Reviewer: Is the addition of this line on purpose?

Author: No 😅

# Documentation

DataShuttle is a tool to streamline the management and standardisation of neuroscience project folders and files.

> DataShuttle's goal is to alleviate the burden researchers face in adhering to standardized file and folder specifications during the execution of intricate and demanding experimental projects. It will:

Reviewer 2: amaerican spelling of "standardized". British spelling used above

- Eliminate the need to manually integrate datasets collected across different machines (e.g. behaviour and electrophysiology acquisition machines).
- Allow convenient transfer of data between machines. This may be between a central project storage and analysis machine (e.g. ''*I want to transfer subjects 1-5, sessions 5 and 10, behavioural data only to my laptop*.'')
- Avoids re-naming and re-formatting of project folders for collaboration or dataset publication.

> DataShuttle aims to integrate seamlessly into the  neuroscience data collection and analysis workflows and eliminate the need to manually , providing tools to:

Reviewer 3: ~and eliminate the need to manually~ ?

- Create folder trees that adhere to SWC-Blueprint, a data management specification based on and aligned to the Brain Imaging Dataset Specification (BIDS), widely used in neuroscience.

> - Convenient transfer of between machines used for data collection and analysis, and a central storage repository.

Reviewer 3: Allow convenient transfer of data? (but this is also in goals)

[IMAGE OF PCS]
[TODO] - make clear that backlash means newline in CLI (not always clear in guides)

DataShuttle requires a one-time setup of project name and configurations.  Next, subjects, session and data-type folder trees can be conveniently created during experimental acquisition. Once acquisition is complete, data can be easily transferred from acquisition computers to a central storage machine.

### Installation

DataShuttle is hosted on  [PyPI](https://pypi.org/project/datashuttle/) and can be installed with pip.

`pip install datashuttle`

Datashuttle additionally requires Rclone for data transfers. The easiest way to install Rclone is using [Miniconda](https://docs.conda.io/en/main/miniconda.html):

```
conda install -c conda-forge rclone
```

See [the Rclone website](https://rclone.org/install/) for alternative installation methods.

## Getting Started

> Datashuttle provides a Python API and cross-platform command line interface (CLI). In this guide examples will be down using the command line, but corresponding methods can be found in the [API Reference](https://datashuttle.neuroinformatics.dev/pages/api_index.html).

Reviewer 3: examples will be *shown*

The first thing to do when using DataShuttle is to setup a new project on a *local* machine.

#### *local* machines and the *central* machine

DataShuttle makes the distinction between (possibly multiple) *local* machines and a single *central* machine. DataShuttle needs to be setup once for each *local* machine, but requires no setup on the *central* machine.

A typical use case is an experiment in which behavioural data and electrophysiological data are collected on acquisition PCs. They send the data to a central server where it is stored.

Later, a subset of the data is transferred to a third machine for analysis. In this case, the behavioural and electrophysiological acquisition machine and analysis machines are 'local'. The central storage machine is the *central* machine.

### One-time project setup on a *local* machine

A one-time setup on each *local* machine used is required, specifying the *project name* and *configurations*.

The configurations tell DataShuttle

- The paths to the *local* and *central* folders that contain the project.
- How to connect to the central project.
- The settings that specify how data is transferred.

> - The *data-types* (e.g. *behaviour* (`behav`), *electrophysiology* (`ephys`)) that will be used in the project.

Reviewer 2: To avoid double parentheses:
```suggestion
- The *data-types* that will be used in the project, e.g. *behaviour* (`behav`) or *electrophysiology* (`ephys`)
```

The command `make-config-file` is used for the initial setup of the project. The **required arguments** are:

`local_path`: The full filepath to the project folder on the *local* machine. For example, if you wanted to make a new project called `my_first_project` in the folder `C:\User\my_projects`, the local path would be `C:\User\my_projects`.

`central_path`: The path on the *central* machine to the central project. For example, if connecting to a remote linux server, this may be `/hpc/home/user/my_projects`.

`connection_method`: `local_filesystem` or `ssh`. Local filesystem can be used if the *central* storage is mounted to the local machine. Otherwise `ssh` can be used.

Finally, the *data-type* flags `--use_ephys`, `--use_funcimg`, `--use_histology`, `--use_behav` set the types of data required for the project on the local machine. While individual flags are optional, at least one must be chosen when initialising the project.

**Optional Arguments**

> If connection method is `ssh`, the `central_host_id`, `central_host_username` must be set, and a one-time SSH setup command run (see the [SSH section][#### SSH] for details).

Reviewer 3: broken section ref

> The optional arguments `ovewrite_old_files`, `transfer_verbosity` and `show_transfer_progress` determine how *data transfer* is performed (see the [Data Transfer section](#### Data Transfer) for details).

Reviewer 3: broken section ref

Reviewer 3: overwrite typo

**Example**

All examples are given using the command-line interface (see the [TODO API] section for identical Python API command). Note that in the terminal, `\` indicates a new-line (allowing a single command to be spread across multiple lines for display purposes). [TODO can somone check if this is the case on windows].

> An example call to `make-config-file` below sets makes a new project called `my_first_project`, sets the *local* project path to `/path/to/my/project`, the *central* path (to a remote Linux server) to `/nfs/nhome/live/username/`, sets the required SSH configurations, and indicates that *behavioural*, electrophysiological and *histological* data will be used on this machine for this project.

Reviewer 2: Two verbs "sets makes" back-to-back

DataShuttle offers a convenient way of transferring entire project folders, or subsets of data. For example, the call

```
datashuttle \
my_first_project \
upload
-sub 001@TO@003
-ses 005_date-@*@ 006_date-@*@*
-dt behav
```

Will *upload* (from *local* to *central* ) behavioural sessions 5 and 6, collected at any date, for subjects 1 to 3.

The keyword `all` can be input in place of a -`sub`, `-ses` or data type argument `dt` to transfer all available subject, sessions or data types available. For example:

```
datashuttle \
my_first_project \
download
-sub all
-ses 005
-dt ephys
```
Will transfer *electrophysiology* data of the 5th sessions for *all* subjects found on the *central* machine.

The *download* command transfers data from the *central* to *local* PC. This can be useful in case you want to analyse a subset of data that is held in *central* storage.

> The main data transfer commands are: `upload`, `download`, `upload-working-folder`, `download-working-folder`, `upload-entire-project`, `download-entire-project`. To understand their behaviour, it is necessary to understand the concept of the *top level folder*.

Reviewer 2: Add link to the relevant section about top level folder

If you want to quickly transfer an entire project (i.e. all data in both `rawdata` and `derivatives`), the command `upload-entire-project` or `download-entire-project` can be used.

e.g. the command
`datashuttle my_first_project upload-entire-project`

run on the folder tree:

```
.
└── my_first_project/
    ├── rawdata/
    │   └── sub-001/
    │       └── ses-001/
    │           └── my_tracking_video.mp4
    └── derivatives/
        └── sub-001/
            └── tracking_video_results.csv
```

will transfer all data in both the `rawdata` and `derivatives` folders from the *local* machine to the *central* machine.

### Summary

This concludes the *Get Started* part of the documentation. Hopefully, you are now equipped to manage an experimental project folder aligned to community standards with a few short commands. Say goodbye to the days of manual copying and pasting!

Continue reading the documentation for a full overview of DataShuttle functionalities. This includes XXX, XXX, XXX.

To discuss, contribute or give feedback on DataShuttle, please check out our discussions page and GitHub repository. Any feedback is welcomed and greatly appreciated!

> ## Documentation

Reviewer 2: Maybe this subsection can be renamed to "More info" or something similar?

Reviewer 2: Or perhaps "Advanced usage"

	transfer_verbosity="v",
	use_ephys=True,
	use_behav=True,
	use_histology=True,
)
```

and methods for making subject folders and transferring data accessed similarly. Note that the shortcut arguments `-sub`, `-ses`, `-dt` are not available through the Python API, and the full argument names (`sub_names`, `ses_names`, `data_type`) must be used.

```
project.make_sub_folders(
	sub_names="sub-001@TO@002",
	ses_names="ses-001_@DATE@",
	data_type="all"
)
```

```
project.upload(
	sub_names="001@TO@003",
	ses_names=["005_date-@*@", "006_date-@*@"],
	data_type="behav"
)
```

### Data Types

[TODO] Link to SWC-Blueprint here, no need to re-invent the wheel. Mention how `histology` is subject-level but all others are `session`. This is actually easy to re-configure by adapting a value in the source code, but it (currently) is not exposed.

### Setting up the connection to *central*

#### Local Filesystem

Local filesystem transfers are typically used when the *central* machine is setup as a mounted drive. This is a common form of communication between client machines and a central server, such as a high-performance computer (HPC, also often called *clusters*).

> When a *central* machine is mounted to the *Local* machine, it acts as it is available as a local-filesystem folder. In this case, the `central_path` configuration (see `make_config_file`) can simply be set to the path directing to the mounted drive.  [TODO: example]

Reviewer 2: it acts as **if**

#### Overwriting existing files

The most important optional configuration, `overwrite_old_files` determines whether folders and files are overwritten during transfer. By default, DataShuttle does not overwrite during data transfer.

For example, if the file `sub-001_ses-001_measure-trajectories.csv` has been transferred from *local* to the *remote* repository, it will never be over-written in the *remote* repository. If the file is edited locally, such that the timestamp on the file is more recent in *local* compared to *remote*, the *remote* version will still not be overwritten.

To change this behaviour, the configuration `overwrite_old_files` can be set to `True`. In this case, files in which the timestamp of the target directory (e.g. *central* when performing `upload`) will be overwritten if their timestamp is older than the corresponding file in the source directory (e.g. *local* when performing `upload`.
)

#### Additional Transfer Configurations

> `transfer_verbosity` : set to `"vv"` for a extensive detail on the transfer operation. Set to `"v"` to see simply each file that is transferred as well as significant events that occur during transfer.

Reviewer 2: ```suggestion
`transfer_verbosity` : set to `"vv"` for a extensive detail on the transfer operation. Set to `"v"` to simply see each file that is transferred as well as significant events that occur during transfer.
```

`show_transfer_progress` : When `True`, real-time transfer statistics will be reported and logged.

> #### Specifying transfers with keyword arguments [TODO better name]

Reviewer 2: I found this section quite confusing. The commands and their effects are not really intuitive. We should plan to rethink these funcitonalities at some point. We could discuss this during the team meeting presentation.

DataShuttle provides a number of keyword arguments to allow separate handling of files that are not found in *data-type* folders.
**For use with the `-sub` / `--sub-names` flag**

`all` : All *subject* and non-*subject* files and folders within the *top level folder* (e.g. `rawdata`) will be transferred.

> `all_sub` : *Subject*  <u>folders</u> only (i.e. prefixed with `-sub)` will be transferred.

Reviewer 2: ```suggestion
`all_sub` : *Subject*  <u>folders</u> only (i.e. prefixed with `-sub`) will be transferred.
```

`all_non_sub` : All files and folders that are not prefixed with `-sub` will be transferred. Any folders prefixed with `-sub` will not be transferred.

**For use with the `-ses` / `--ses-names` flag**

`all` : All *session* and non-*session* files and folders within a *subject* level folder (e.g. `sub-001`) will be transferred.

> `all_sub` : *Session* <u>folders</u> only (i.e. prefixed with `-ses`) will be transferred. Note that the only exception is the `histology` folder, the transfer of which is determined by the `-dt` flag (below).

Reviewer 2: should be `all_ses`, right?

- All *data-type* folder contents (`histology`, `behav`, `ephys`)

The command:

```
datashuttle \
my_first_project \
upload
-sub 001
-ses all_sub
-dt all_data_type
```

Would upload:

- Contents residing in the `sub-001` folder only.
- All *data-type* folder contents (`histology`, `behav`, `ephys`)

3) The final example shows the effect of transferring `all_non_sub` files only. The command:

```
datashuttle \
my_first_project \
upload
-sub all_non_sub
-ses all
-dt all
```

Would upload:

- the file `a_project_related_file.json` only.

#### Query DataShuttle for current *configurations*

It is possible to query DataShuttle for the current *configurations* and relevant file paths,
for example the `show-local-path` command will print the currently set *local* path to the terminal.

Similarly, the command `show-configs` will print all currently set *configurations*. A number of additional convenience functions exist (for example `show-remote-path`, `show-top-level-folder`, `show-config-path`). For a full list, see our API reference [TODO: link].

### Logging

Detailed logs of all configuration changes, folder creation and data transfers are logged

> to the `.datashuttle` folder in the that is created in the *local* project folder.

Reviewer 2: ```suggestion
to the `.datashuttle` folder that is created in the *local* project folder.
```

For each command run, a log of that command is placed in the logs folder, with the time and date of the command. The log itself contains relevant information pertaining to that command. For example, if the commands `make_sub_folders`, `upload`, `download` were run sequentially, the logs output folder would look like:

```
make_sub_folders_2023-06-08_09-55-14.log
upload_data_2023-06-08_09-55-45.log

> download_data_2023-06-08_09-56-19.log

Reviewer 2: Idea for the future. Format log file names as <data_time_command.log>. This way they will be displayed chnonologically in the folder, when alphabetically sorted
